### PR TITLE
Functions added in client socket

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: internal/controller/websocket.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import (
+	"Tradeasy/internal/services/stock_exchange"
+	"encoding/json"
+	"github.com/gin-gonic/gin"
+	"github.com/gorilla/websocket"
+	"log"
+	"time"
+)
+
+//var OrderUpdated = make(chan stock_exchange.OrderResponse)
```
Comment (mukesh0513) [inline internal/controller/websocket.go]: remove.

File: internal/controller/websocket.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	var upgrader = websocket.Upgrader{} // use default options
+	// Upgrade our raw HTTP connection to a websocket based one
+	conn, err := upgrader.Upgrade(c.Writer, c.Request, nil)
+	if err != nil {
+		log.Print("Error during connection upgradation:", err)
+		return
+	}
+	defer conn.Close()
+	//write stock details
+	writeEvery(time.Duration(5)*time.Millisecond*1000, conn, stock_exchange.StockWrite)
```
Comment (mukesh0513) [inline internal/controller/websocket.go]: Define this time as constant or in configuration file.

File: internal/controller/websocket.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	var upgrader = websocket.Upgrader{} // use default options
+	// Upgrade our raw HTTP connection to a websocket based one
+	conn, err := upgrader.Upgrade(c.Writer, c.Request, nil)
+	if err != nil {
+		log.Print("Error during connection upgradation:", err)
+		return
+	}
+	defer conn.Close()
+
+	/*go func() {
```
Comment (mukesh0513) [inline internal/controller/websocket.go]: Remove if not needed.

File: internal/controller/websocket.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+		for {
+			time.Sleep(5 * time.Second)
+			OrderUpdated <- stock_exchange.OrderResponse{AveragePrice: uint(rand.Intn(10))}
+		}
+	}()*/
+
+	for {
```
Comment (mukesh0513) [inline internal/controller/websocket.go]: Where is OrderUpdated defined and what is scope of it?

File: internal/controller/websocket.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+			if err := conn.WriteMessage(websocket.TextMessage, orderJson); err != nil {
+				log.Println("Error during writing stocks to websocket:", err)
+				return
+			}
+		}
+	}
+}
+
+func writeEvery(d time.Duration, conn *websocket.Conn, f func() (*[]stock_exchange.StockDetails, error)) {
```
Comment (mukesh0513) [inline internal/controller/websocket.go]: do we need to pass function handler in arguments?

File: internal/services/websocket.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
+package services
```
Comment (mukesh0513) [inline internal/services/websocket.go]: Move this into right package folder.

File: internal/services/websocket.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+		var orderDetails stock_exchange.OrderResponse
+		err=json.Unmarshal(orderMessage,&orderDetails)
+		if err!=nil{
+			log.Println("Error during Unmarshalling:",err)
+			break
+		}
+		var p model.PendingOrders
+		if err=config.DB.Table("pending_orders").Where("order_id=?",orderDetails.OrderID).First(&p).Error;err!=nil{
+			log.Println("Order_id doesnt match with pending_orders table:",err)
+			break
```
Comment (mukesh0513) [inline internal/services/websocket.go]: breaking will stop reading right?

File: internal/services/websocket.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+		_, orderMessage, err := conn.ReadMessage()
+		if err != nil {
+			log.Println("Error during message reading:", err)
+			break
+		}
+		var orderDetails stock_exchange.OrderResponse
+		err=json.Unmarshal(orderMessage,&orderDetails)
+		if err!=nil{
+			log.Println("Error during Unmarshalling:",err)
+			break
```
Comment (mukesh0513) [inline internal/services/websocket.go]: [Comment](https://github.com/RohanDudeja/Tradeasy/pull/52/files#r685027863)

File: internal/services/websocket.go (hunk 7 above)
Comment (mukesh0513) [reply internal/services/websocket.go]: Check this all over the for loop.

File: internal/services/websocket.go [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+		}else if p.OrderType=="Sell"{
+			err=order.UpdateSellOrder(orderDetails)
+			if err!=nil {
+				log.Println("Error in updating buy order")
+				break
+			}
+		}
+
+
+		log.Printf("Received: %s", orderMessage)
```
Comment (mukesh0513) [inline internal/services/websocket.go]: Write descriptive logging message.

File: internal/services/websocket.go [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+	stockConn, _, err := websocket.DefaultDialer.Dial(socketUrl, nil)
+
+	if err != nil {
+		log.Fatal("Error connecting to Websocket Server:", err)
+	}
+	defer stockConn.Close()
+
+	// Our main loop
+	for {
+		_, stockMessage, err := stockConn.ReadMessage()
```
Comment (mukesh0513) [inline internal/services/websocket.go]: Same comments as of orderConnection.

File: internal/controller/websocket.go (hunk 4 above)
Comment (RohanDudeja) [reply internal/controller/websocket.go]: OrderUpdated is defined in services for stock exchange engine

File: internal/controller/websocket.go (hunk 5 above)
Comment (RohanDudeja) [reply internal/controller/websocket.go]: moved the content of function to StockHandler itself

File: internal/services/websocket.go (hunk 7 above)
Comment (ghost) [reply internal/services/websocket.go]: Added continue in place of break.

File: internal/services/websocket.go (hunk 6 above)
Comment (ghost) [reply internal/services/websocket.go]: moved to the order folder in service.

File: internal/controller/websocket.go [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+import (
+	"Tradeasy/internal/services/stock_exchange"
+	"encoding/json"
+	"github.com/gin-gonic/gin"
+	"github.com/gorilla/websocket"
+	"log"
+	"time"
+)
+
+const TimeInterval = time.Duration(5)*time.Millisecond*1000
```
Comment (mukesh0513) [inline internal/controller/websocket.go]: use time.seconds directly.

Comment (mukesh0513) [reply internal/controller/websocket.go]: Name variable as for what purpose is this being used.

File: internal/services/stock_exchange/stock_exchange.go [hunk 12]
```diff
@@ -1,5 +1,12 @@
 package stock_exchange
 
+import (
+	"Tradeasy/config"
+	"log"
+)
+
+var OrderUpdated = make(chan OrderResponse)
```
Comment (mukesh0513) [inline internal/services/stock_exchange/stock_exchange.go]: can we add some comments here on how is this being used?

File: internal/controller/websocket.go (hunk 4 above)
Comment (mukesh0513) [reply internal/controller/websocket.go]: Are you initialising OrderUpdated globally? or do you want to keep its scope restricted.

File: internal/controller/websocket.go (hunk 11 above)
Comment (RohanDudeja) [reply internal/controller/websocket.go]: renamed to stockTimeInterval and added time.second

File: internal/services/order/websocket.go [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+			log.Println("Error during Unmarshalling:",err)
+			continue
+		}
+		go func() {
+			err=UpdateStocksFeed(stockDetails)
+		}()
+		if err!=nil {
+			log.Println("Error during saving the stocks feed:",err)
+			continue
+		}
```
Comment (mukesh0513) [inline internal/services/order/websocket.go]: redundant.

**Review by mukesh0513 — APPROVED** [review]
LGTM

**Pull request merged**